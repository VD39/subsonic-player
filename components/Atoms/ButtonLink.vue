<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
