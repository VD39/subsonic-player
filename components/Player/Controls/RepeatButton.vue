<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { repeat, setRepeat } = useAudioPlayer();

const buttonProps = computed<RepeatButtonProps>(() => {
  const noRepeat = repeat.value === -1;

  return {
    icon: repeat.value === 1 ? 'PhRepeatOnce' : 'PhRepeat',
    iconColor: noRepeat ? 'currentColor' : 'var(--theme-color)',
    iconWeight: noRepeat ? 'duotone' : 'fill',
  };
});

const title = computed(() => {
  switch (repeat.value) {
    case -1:
      return 'Turn on repeat all.';
    case Number.POSITIVE_INFINITY:
      return 'Turn on repeat one.';
    default:
      return 'Turn repeat off.';
  }
});
</script>

<template>
  <IconButton
    :icon="buttonProps.icon"
    :icon-color="buttonProps.iconColor"
    :icon-weight="buttonProps.iconWeight"
    :title="title"
    @click="setRepeat"
  >
    {{ title }}
  </IconButton>
</template>
