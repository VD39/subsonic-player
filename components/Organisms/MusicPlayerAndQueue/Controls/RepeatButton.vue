<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const { repeat, setRepeat } = useAudioPlayer();

const buttonProps = computed<ButtonProps>(() => {
  const noRepeat = repeat.value === -1;

  return {
    icon: repeat.value === 1 ? ICONS.repeatOnce : ICONS.repeat,
    iconColor: noRepeat ? 'currentColor' : 'var(--theme-color)',
    iconWeight: noRepeat ? 'regular' : 'fill',
  };
});

const title = computed(() => {
  switch (repeat.value) {
    case -1:
      return 'Turn on repeat all';
    case Number.POSITIVE_INFINITY:
      return 'Turn on repeat one';
    default:
      return 'Turn repeat off';
  }
});
</script>

<template>
  <ButtonLink
    :icon="buttonProps.icon"
    :iconColor="buttonProps.iconColor"
    :iconWeight="buttonProps.iconWeight"
    :title
    @click="setRepeat"
  >
    {{ title }}
  </ButtonLink>
</template>
