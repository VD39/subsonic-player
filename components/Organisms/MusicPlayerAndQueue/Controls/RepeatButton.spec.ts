import type { VueWrapper } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { mount } from '@vue/test-utils';

import RepeatButton from './RepeatButton.vue';

const { repeatMock, setRepeatMock } = useAudioPlayerMock();

function factory(props = {}) {
  return mount(RepeatButton, {
    props: {
      ...props,
    },
  });
}

const buttonProps = {
  all: {
    color: 'var(--theme-color)',
    icon: ICONS.repeat,
    text: 'Turn on repeat one',
    weight: 'fill',
  },
  off: {
    color: 'currentColor',
    icon: ICONS.repeat,
    text: 'Turn on repeat all',
    weight: 'regular',
  },
  one: {
    color: 'var(--theme-color)',
    icon: ICONS.repeatOnce,
    text: 'Turn repeat off',
    weight: 'fill',
  },
};

describe('RepeatButton', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  describe.each([
    [-1, buttonProps.off],
    [1, buttonProps.one],
    [Number.POSITIVE_INFINITY, buttonProps.all],
  ])('when shuffle is set to %s', (repeat, buttonProps) => {
    beforeEach(() => {
      repeatMock.value = repeat;
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct icon on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).props('icon')).toBe(
        buttonProps.icon,
      );
    });

    it('sets the correct color on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).props('iconColor')).toBe(
        buttonProps.color,
      );
    });

    it('sets the correct weight on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).props('iconWeight')).toBe(
        buttonProps.weight,
      );
    });

    it('sets the correct title attribute on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).attributes('title')).toBe(
        buttonProps.text,
      );
    });

    it('sets the correct slot data on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).text()).toContain(
        buttonProps.text,
      );
    });
  });

  describe('when the ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent(ButtonLink).vm.$emit('click');
    });

    it('calls the setRepeat function', () => {
      expect(setRepeatMock).toHaveBeenCalled();
    });
  });
});
