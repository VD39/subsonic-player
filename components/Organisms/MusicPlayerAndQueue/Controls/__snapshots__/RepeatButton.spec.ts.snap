// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RepeatButton > when shuffle is set to -1 > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Turn on repeat all">
  <phrepeat class="icon" aria-hidden="true" size="24" weight="regular" color="currentColor"></phrepeat><span class="clamp text visuallyHidden">Turn on repeat all</span>
</button>"
`;

exports[`RepeatButton > when shuffle is set to 1 > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Turn repeat off">
  <phrepeatonce class="icon" aria-hidden="true" size="24" weight="fill" color="var(--theme-color)"></phrepeatonce><span class="clamp text visuallyHidden">Turn repeat off</span>
</button>"
`;

exports[`RepeatButton > when shuffle is set to Infinity > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Turn on repeat one">
  <phrepeat class="icon" aria-hidden="true" size="24" weight="fill" color="var(--theme-color)"></phrepeat><span class="clamp text visuallyHidden">Turn on repeat one</span>
</button>"
`;
